### DISCUSSION TITLE
Error downloading file .github/technolinator.yml for repository {user}/{Repo}

### DISCUSSION THREAD

**Discussion by smu1990** [MediaMarktSaturn/technolinator#394] [Q&A] [answered]
Hi Team,

I need a help from you guys to find the fix for below error.

Exception: java.lang.IllegalStateException: Error downloading file .github/technolinator.yml for repository {user}/{repo}
        at io.quarkiverse.githubapp.runtime.github.GitHubFileDownloader.getFileContent(GitHubFileDownloader.java:40)
        at io.quarkiverse.githubapp.runtime.github.GitHubFileDownloader_ClientProxy.getFileContent(Unknown Source)
        at io.quarkiverse.githubapp.runtime.github.GitHubConfigFileProviderImpl.fetchConfigFile(GitHubConfigFileProviderImpl.java:58)
        at io.quarkiverse.githubapp.runtime.github.GitHubConfigFileProviderImpl.fetchConfigFile(GitHubConfigFileProviderImpl.java:48)
        at io.quarkiverse.githubapp.runtime.github.GitHubConfigFileProviderImpl_ClientProxy.fetchConfigFile(Unknown Source)
        at io.quarkiverse.githubapp.runtime.RequestScopeCachingGitHubConfigFileProvider.lambda$getConfigObject$0(RequestScopeCachingGitHubConfigFileProvider.java:36)
        at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)     
        at io.quarkiverse.githubapp.runtime.RequestScopeCachingGitHubConfigFileProvider.getConfigObject(RequestScopeCachingGitHubConfigFileProvider.java:35)
        at io.quarkiverse.githubapp.runtime.RequestScopeCachingGitHubConfigFileProvider_ClientProxy.getConfigObject(Unknown Source)
        at com.mediamarktsaturn.technolinator.events.OnPullRequestDispatcher_Multiplexer.onPullRequest_655d6992b64c58f9eecfa2da12bf390cd0c7ac59(Unknown Source)
        at com.mediamarktsaturn.technolinator.events.OnPullRequestDispatcher_Multiplexer_Observer_onPullRequest_655d6992b64c58f9eecfa2da12bf390cd0c7ac59_1vUmpQWjlYEhwrtSsOfbG008qjw.notify(Unknown Source)
        at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:346)
        at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:328)
        at io.quarkus.arc.impl.EventImpl$1.get(EventImpl.java:112)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)     
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:582)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.kohsuke.github.HttpException: {"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/repos/contents#get-repository-content"}

I am getting above error while running technolinator on Dev mode. 

I tried also from the docker-compose which is packed with DTfronend and DTapi server etc.. 

{
	"timestamp": "2024-03-15T18:42:10.090498807Z",
	"sequence": 67,
	"loggerClassName": "org.jboss.logging.Logger",
	"loggerName": "io.quarkiverse.githubapp",
	"level": "ERROR",
	"message": "Error handling delivery ba0df690-e2fb-11ee-90df-b2f6fcbff519\n› Repository: {user}/{repo}\n› Event:      pull_request.opened\n› Context:    https://github.com/smu1990/OSLC_Consumer_Code-master/pull/27\n› Redeliver:  https://github.com/settings/apps/Technolinator/advanced\nException",
	"threadName": "executor-thread-2",
	"threadId": 58,
	"mdc": {},
	"ndc": "",
	"hostName": "c5a641479440",
	"processName": "quarkus-run.jar",
	"processId": 1,
	"exception": {
		"refId": 1,
		"exceptionType": "java.lang.IllegalStateException",
		"message": "Error downloading file .github/technolinator.yml for repository smu1990/OSLC_Consumer_Code-master",
		"frames": [
			{
				"class": "io.quarkiverse.githubapp.runtime.github.GitHubFileDownloader",
				"method": "getFileContent",
				"line": 40
			},
			{
				"class": "io.quarkiverse.githubapp.runtime.github.GitHubFileDownloader_ClientProxy",
				"method": "getFileContent"
			},
			{
				"class": "io.quarkiverse.githubapp.runtime.github.GitHubConfigFileProviderImpl",
				"method": "fetchConfigFile",
				"line": 58
			},
			{
				"class": "io.quarkiverse.githubapp.runtime.github.GitHubConfigFileProviderImpl",
				"method": "fetchConfigFile",
				"line": 48
			},
			{
				"class": "io.quarkiverse.githubapp.runtime.github.GitHubConfigFileProviderImpl_ClientProxy",
				"method": "fetchConfigFile"
			},
			{
				"class": "io.quarkiverse.githubapp.runtime.RequestScopeCachingGitHubConfigFileProvider",
				"method": "lambda$getConfigObject$0",
				"line": 36
			},
			{
				"class": "java.util.concurrent.ConcurrentHashMap",
				"method": "computeIfAbsent",
				"line": 1708
			},
			{
				"class": "io.quarkiverse.githubapp.runtime.RequestScopeCachingGitHubConfigFileProvider",
				"method": "getConfigObject",
				"line": 35
			},
			{
				"class": "io.quarkiverse.githubapp.runtime.RequestScopeCachingGitHubConfigFileProvider_ClientProxy",
				"method": "getConfigObject"
			},
			{
				"class": "com.mediamarktsaturn.technolinator.events.OnPullRequestDispatcher_Multiplexer",
				"method": "onPullRequest_655d6992b64c58f9eecfa2da12bf390cd0c7ac59"
			},
			{
				"class": "com.mediamarktsaturn.technolinator.events.OnPullRequestDispatcher_Multiplexer_Observer_onPullRequest_655d6992b64c58f9eecfa2da12bf390cd0c7ac59_1vUmpQWjlYEhwrtSsOfbG008qjw",
				"method": "notify"
			},
			{
				"class": "io.quarkus.arc.impl.EventImpl$Notifier",
				"method": "notifyObservers",
				"line": 346
			},
			{
				"class": "io.quarkus.arc.impl.EventImpl$Notifier",
				"method": "notify",
				"line": 328
			},
			{
				"class": "io.quarkus.arc.impl.EventImpl$1",
				"method": "get",
				"line": 112
			},
			{
				"class": "java.util.concurrent.CompletableFuture$AsyncSupply",
				"method": "run",
				"line": 1768
			},
			{
				"class": "io.quarkus.vertx.core.runtime.VertxCoreRecorder$14",
				"method": "runWith",
				"line": 582
			},
			{
				"class": "org.jboss.threads.EnhancedQueueExecutor$Task",
				"method": "run",
				"line": 2513
			},
			{
				"class": "org.jboss.threads.EnhancedQueueExecutor$ThreadBody",
				"method": "run",
				"line": 1512
			},
			{
				"class": "org.jboss.threads.DelegatingRunnable",
				"method": "run",
				"line": 29
			},
			{
				"class": "org.jboss.threads.ThreadLocalResettingRunnable",
				"method": "run",
				"line": 29
			},
			{
				"class": "io.netty.util.concurrent.FastThreadLocalRunnable",
				"method": "run",
				"line": 30
			},
			{
				"class": "java.lang.Thread",
				"method": "run",
				"line": 1583
			}
		],
		"causedBy": {
			"exception": {
				"refId": 2,
				"exceptionType": "org.kohsuke.github.HttpException",
				"message": "{\"message\":\"Resource not accessible by integration\",\"documentation_url\":\"https://docs.github.com/rest/repos/contents#get-repository-content\"}",
				"frames": [
					{
						"class": "org.kohsuke.github.GitHubConnectorResponseErrorHandler$1",
						"method": "onError",
						"line": 62
					},
					{
						"class": "org.kohsuke.github.GitHubClient",
						"method": "detectKnownErrors",
						"line": 504
					},
					{
						"class": "org.kohsuke.github.GitHubClient",
						"method": "sendRequest",
						"line": 464
					},
					{
						"class": "org.kohsuke.github.GitHubClient",
						"method": "sendRequest",
						"line": 427
					},
					{
						"class": "org.kohsuke.github.Requester",
						"method": "fetch",
						"line": 85
					},
					{
						"class": "org.kohsuke.github.GHRepository",
						"method": "getFileContent",
						"line": 2712
					},
					{
						"class": "io.quarkiverse.githubapp.runtime.github.GitHubFileDownloader",
						"method": "getFileContent",
						"line": 27
					},
					{
						"class": "io.quarkiverse.githubapp.runtime.github.GitHubFileDownloader_ClientProxy",
						"method": "getFileContent"
					},
					{
						"class": "io.quarkiverse.githubapp.runtime.github.GitHubConfigFileProviderImpl",
						"method": "fetchConfigFile",
						"line": 58
					},
					{
						"class": "io.quarkiverse.githubapp.runtime.github.GitHubConfigFileProviderImpl",
						"method": "fetchConfigFile",
						"line": 48
					},
					{
						"class": "io.quarkiverse.githubapp.runtime.github.GitHubConfigFileProviderImpl_ClientProxy",
						"method": "fetchConfigFile"
					},
					{
						"class": "io.quarkiverse.githubapp.runtime.RequestScopeCachingGitHubConfigFileProvider",
						"method": "lambda$getConfigObject$0",
						"line": 36
					},
					{
						"class": "java.util.concurrent.ConcurrentHashMap",
						"method": "computeIfAbsent",
						"line": 1708
					},
					{
						"class": "io.quarkiverse.githubapp.runtime.RequestScopeCachingGitHubConfigFileProvider",
						"method": "getConfigObject",
						"line": 35
					},
					{
						"class": "io.quarkiverse.githubapp.runtime.RequestScopeCachingGitHubConfigFileProvider_ClientProxy",
						"method": "getConfigObject"
					},
					{
						"class": "com.mediamarktsaturn.technolinator.events.OnPullRequestDispatcher_Multiplexer",
						"method": "onPullRequest_655d6992b64c58f9eecfa2da12bf390cd0c7ac59"
					},
					{
						"class": "com.mediamarktsaturn.technolinator.events.OnPullRequestDispatcher_Multiplexer_Observer_onPullRequest_655d6992b64c58f9eecfa2da12bf390cd0c7ac59_1vUmpQWjlYEhwrtSsOfbG008qjw",
						"method": "notify"
					},
					{
						"class": "io.quarkus.arc.impl.EventImpl$Notifier",
						"method": "notifyObservers",
						"line": 346
					},
					{
						"class": "io.quarkus.arc.impl.EventImpl$Notifier",
						"method": "notify",
						"line": 328
					},
					{
						"class": "io.quarkus.arc.impl.EventImpl$1",
						"method": "get",
						"line": 112
					},
					{
						"class": "java.util.concurrent.CompletableFuture$AsyncSupply",
						"method": "run",
						"line": 1768
					},
					{
						"class": "io.quarkus.vertx.core.runtime.VertxCoreRecorder$14",
						"method": "runWith",
						"line": 582
					},
					{
						"class": "org.jboss.threads.EnhancedQueueExecutor$Task",
						"method": "run",
						"line": 2513
					},
					{
						"class": "org.jboss.threads.EnhancedQueueExecutor$ThreadBody",
						"method": "run",
						"line": 1512
					},
					{
						"class": "org.jboss.threads.DelegatingRunnable",
						"method": "run",
						"line": 29
					},
					{
						"class": "org.jboss.threads.ThreadLocalResettingRunnable",
						"method": "run",
						"line": 29
					},
					{
						"class": "io.netty.util.concurrent.FastThreadLocalRunnable",
						"method": "run",
						"line": 30
					},
					{
						"class": "java.lang.Thread",
						"method": "run",
						"line": 1583
					}
				]
			}
		}
	}
}

Can someone explain me what could be the issue.?

Thanks in Advance,
Manju

**Comment by heubeck** [comment]
did you provide all of the required permissions to your app installation?

https://github.com/MediaMarktSaturn/technolinator/blob/main/docs/Adoption.md#permissions

haven't seen this error before, but from my gut feeling, it looks like maybe the _content_ permission is missing?

**Reply by smu1990** [reply]
Hi @heubeck ,

thanks for your quick response.

I have provided all the permissions which is mentioned in the Adoption.md document.

![image](https://github.com/MediaMarktSaturn/technolinator/assets/17971827/0be1fcb1-48e7-4b4d-a00c-0e2e5129c81c)

Thanks again

**Reply by heubeck** [reply]
great, thx for the trust, but there's no reason for contents _write_ ;)

hmm... stumbled over the `Expiration time' claim ('exp') must be a numeric value representing the future time at which the assertion expires` in your stacktrace.
is there maybe a clock screw on the machine running technolinator?

edit: if that's not the case, can you provide a rough overview on your setup?

**Reply by smu1990** [reply]
Yes, you are right. 
My server machine date was out of sync. After the correction i am getting the same error even in Docker container technolinator.
i just modified the question.

**Reply by heubeck** [reply, marked as answer]
ok. now we're with error `Resource not accessible by integration`.
did you update the permissions after the installation of the app to the repo and maybe have to re-approve the installation for the repo?

otherwise, just for the sake of testing, would you please create a `.github/technolinator.yml` in the repo?
it has to work without (and it does), but maybe we're having some yet unknown constellation here.

**Reply by smu1990** [reply]
Step 1: Modified docker-compose.yml with localhost settings.
Step 2: pushed .github/technolinator.yml into my source code repo.
Step 3: Modified sample.env with my privatekey and github token, quakus appid etc.
Step 4: run docker-compose up -d
Step 5: checked with docker ps ( all are up and running )
Step 6: DT app is up and running and imported example SBOM. It shows all vulnerabilities and dependency graph.
Step 7: Created pull request in github.
           At this step i am getting this error.

**Reply by heubeck** [reply]
just a remark, with env `QUARKUS_LOG_CONSOLE_JSON=false` you can make it log plaintext.

**Reply by heubeck** [reply]
are you confident about your github credentials in the technolinator config?
my local .env file looks like
![image](https://github.com/MediaMarktSaturn/technolinator/assets/40993644/e78c0918-109a-424b-9faa-44f4c0d2adb0)

**Reply by heubeck** [reply]
ah damn it, the docker-compose example doesn't hold any technolinator config.
have you provided all mandatory configuration (via env or .env file) from https://github.com/MediaMarktSaturn/technolinator/blob/main/docs/Runtime_Config.md ? (those without defaults)

**Reply by smu1990** [reply]
Thanks a lot for your support and hint.

This was purely permission issue :) Sorry my bad. 

I have not accepted the permissions in the App. After accepting them, Error gone.

Now i can see one more error: cdxgen failed for project "project name" 

"exception": {
		"refId": 1,
		"exceptionType": "java.lang.RuntimeException",
		"message": "java.io.IOException: Cannot run program \"cdxgen\" (in directory \"/tmp/Consumer_Code-master18200443695766295178/smu1990/Consumer_Code-master\"): error=2, No such file or directory",

**Reply by heubeck** [reply]
can you please switch to plain text logging (https://github.com/MediaMarktSaturn/technolinator/discussions/394#discussioncomment-8806506) and provide me the full stacktrace of that error?

edit: there's not any special restriction to your container runtime like no filesystem operations inside container for example?
by default, technolinator operates in `/tmp` for downloading and analyzing the repos, and `/data` is used as the process users HOME, meaning that caches and so on (by the used tools or build systems) are placed there.

**Reply by smu1990** [reply]
Done with logging changes.

I can see in the log that folder creating repocache folder under /tmp. 

I am totally new to this work. Sorry for asking many questions.

2024-03-15 19:41:42,787 INFO  [com.med.tec.os.ProcessCallback$DefaultProcessCallback] (executor-thread-4) [85af2014] Starting 'cdxgen --spec-version 1.5 -o sbom.json --no-recurse --fail-on-error --evidence --project-name OSLC_Consumer_Code-master-OSLC_Consumer_Code-master --no-validate' in /tmp/OSLC_Consumer_Code-master17420389130142392540/src
2024-03-15 19:41:43,517 DEBUG [com.med.tec.os.ProcessCallback$DefaultProcessCallback] (executor-thread-1) [85af2014]: Bom file doesn't exist. Check if cdxgen was invoked with the correct type argument.
2024-03-15 19:41:43,525 INFO  [com.med.tec.os.ProcessCallback$DefaultProcessCallback] (executor-thread-1) [85af2014] succeeded after PT0.738S
2024-03-15 19:41:43,526 INFO  [com.med.tec.han.AnalysisProcessHandler] (executor-thread-1) Analysis of repo smu1990/OSLC_Consumer_Code-master, branch refs/heads/main for 1 projects resulted in 1 x Success with SBOMs 1 x None
2024-03-15 19:41:43,854 INFO  [com.med.tec.han.AnalysisProcessHandler] (executor-thread-1) No pull-requests open in repo https://github.com/smu1990/OSLC_Consumer_Code-master with head main
2024-03-15 19:41:43,857 INFO  [com.med.tec.han.AnalysisProcessHandler] (executor-thread-1) Nothing to analyse in repo https://github.com/smu1990/OSLC_Consumer_Code-master, ref main
2024-03-15 19:41:43,858 INFO  [com.med.tec.os.ProcessCallback$DefaultProcessCallback] (executor-thread-1) [85af2014] Starting 'rm -rf /tmp/OSLC_Consumer_Code-master17420389130142392540' in .
2024-03-15 19:41:43,860 INFO  [com.med.tec.eve.DispatcherBase] (executor-thread-1) Setting repo https://github.com/smu1990/OSLC_Consumer_Code-master commit 8efa5f6dd1f74d966c0a4ed0fbccb8ba4713cd8f status to SUCCESS
2024-03-15 19:41:43,878 INFO  [com.med.tec.os.ProcessCallback$DefaultProcessCallback] (executor-thread-4) [85af2014] succeeded after PT0.02S

It is saying success but in DT no projects got created. Where to find SBOM for this analysis ?

**Reply by heubeck** [reply]
no worries.

ok, from the logs it looks good.
what you always can do, is simple pick up the command it executes (like `cdxgen --spec-version 1.5 -o sbom.json --no-recurse --fail-on-error --evidence --project-name OSLC_Consumer_Code-master-OSLC_Consumer_Code-master --no-validate`) and run it in a local copy of your repository.

the log line `1 projects resulted in 1 x Success with SBOMs 1 x None` state, that the cdxgen command was successful, but no bom (1 None) was created.
what's your test project? is it a source code repository - so anything that could be detected by cdxgen as _component_ or _dependencies_?

the bom itself is transient, you won't see it - that would be a nice debugging option, to not cleanup after analysis (regardless of the result).

**Reply by smu1990** [reply]
Thanks for your hint again.

The command "cdxgen --spec-version 1.5 -o sbom.json --no-recurse --fail-on-error --evidence --project-name OSLC_Consumer_Code-master-OSLC_Consumer_Code-master --no-validate" executed under my local copy and bom.json got created.

I just made "recursive:ture" and project scan is happening from prod profile

SOM also got created as per the log.

![image](https://github.com/MediaMarktSaturn/technolinator/assets/17971827/194122e5-8687-4ece-9462-5c39c2c7499e)

![image](https://github.com/MediaMarktSaturn/technolinator/assets/17971827/1c58c117-8860-4326-9707-39fb465099a9)

Can you please help me what went wrong at DT side.

Thanks

**Reply by heubeck** [reply]
ok, we're getting close.
the "yield" success type states, there was a sbom created, and it even got scored.
I would expect, that the `DTRACK_URL` doesn't match your env, have you set it to the dtrack base host (without `/api`)?

**Reply by smu1990** [reply]
Earlier it was the api key issue which i have corrected in DT admin page.

DTRACK_URL is updated as below.

![image](https://github.com/MediaMarktSaturn/technolinator/assets/17971827/8166570b-4dab-4580-9228-de8403473ed4)

API server is running on 127.0.0.1:8080 port

![image](https://github.com/MediaMarktSaturn/technolinator/assets/17971827/7785aa1b-82ae-4067-b792-e2a90f0deb88)

Now i am getting connection refuse.

![image](https://github.com/MediaMarktSaturn/technolinator/assets/17971827/c8882d6f-444a-4ecf-898c-d383f27b4db3)

But api server is running under above mentioned host and port.

Can you please point me what i am doing wrong here ?

Thanks in advance

**Reply by heubeck** [reply]
from the technolinator containers point of view, localhost is only itself.
please try with
`DTRACK_URL=http://dtrack-apiserver:8080`
that should be accessible within the compose network.

**Reply by smu1990** [reply]
You are awesome :) . you were absolutely right.

Initially i set this up in my windows wsl which was not identifying the api server.

Now moved to linux dedicated machine and complete flow worked. 

Project got created in DT as well. But i don't see any vulnerabilities. 

Below is the complete log

![image](https://github.com/MediaMarktSaturn/technolinator/assets/17971827/c106f48c-0b1b-46f8-9dec-ec2f3a8f868d)

I am scanning this for my JAVA source code.

Am i still doing something wrong ?

**Reply by heubeck** [reply]
no, everything's fine. evidence is only partially supported, even for Java.
and dtrack doesn't uses it at all, so you can also deactivate it for now (`CDXGEN_EVIDENCE_DEFAULT=false` - but is false by default, so just omit the setting).

dtrack analyses projects periodically, you can configure the frequence to a minimum of 1h (dtrack adtmin settings schedule).
so after that time, there should be some metrics for your projects.

**Reply by smu1990** [reply]
Awesome. Thanks for your kind response.

One Last thing which i wanted to ask.

Is there a way that i can trigger dtrack analysis manually ?

**Reply by heubeck** [reply]
hmm, there's a "metric refresh" right under the metric graphs in every project.
but not sure what it exactly does. maybe check the api documentation as well for options.

so your setup is working now?

**Reply by smu1990** [reply]
Yes, my setup is up and running :) Thanks a lot for your guidance.

I little confused on one thing that, if i run cdxgen from my local repo i am getting vulnerabilities in bom.json. 

If i add bom.json into DT, it shows the risks. But  technolinator scans the code but i am not able to see any vulnerability in DT. 

In DT, project got created and also updated some parameters. 

Now i am stuck here :)

**Reply by heubeck** [reply]
yeah, that's a common issue. because every tool has different CVE <-> Component matching logic, there are false-positives or false-negatives around all the time.

you now need to find out, how's right: read the CVE that was discovered and check, if it matches the component which it was discovered from.
most of the tools have a false-positive reporting, like here for depscan: https://github.com/owasp-dep-scan/dep-scan/issues/new/choose
and of course, if the finding is valid, report it to dependency-track, so they can check.

can you share the CVE and the component it was matched to locally?

**Reply by smu1990** [reply]
Sorry it might be a stupid question but since i am new to this. 

Can you please guide me what CVE and where i can find it locally. 

I can see only bom.json and some slice jsons thats all..

**Reply by heubeck** [reply]
sorry, my fault. did assume you already further-on processed the bom.

the bom.json created by cdxgen doesn't contain any vulnerabilities, it's _just_ the ingredients list of the software it was generated from.
the task to check for vulnerable components listed in the bom is up to DT.

to test by your own, or to validate DTs results, you use for instance [depscan](https://github.com/owasp-dep-scan/dep-scan) to analyze your bom: `depscan --bom bom.json`, it will (after some while when started for the first time) output a list of CVEs - or be happy because there's none ;)
that's also what technolinator does for commenting pull-requests.

**Reply by smu1990** [reply]
After lot of changes and adopting your steps, I can able to see everything in DT :) 

This is so beautiful. 

Thanks a lot for your kind support.

**Reply by heubeck** [reply]
you're welcome. hope to read you again soon ;)
please report any issues you experience, propose ideas/improvements and also share your successes :D

**Reply by smu1990** [reply]
Sure, I will be in touch with you :)
